### ISSUE TITLE
Suspicious lack of optimization  in new version

### ISSUE THREAD

**Issue by ddaa2000** [neo4j/neo4j#12926]
I discovered that a query runs more than 1000 times slower in 4.4.10 compared to 4.2.19

**Neo4j Version:** 4.4.10
**Operating System:** Ubuntu 20.04
**API:** java

# Steps to reproduce
1. Execute [graph.txt](https://github.com/neo4j/neo4j/files/9417838/graph.txt) to create the graph
2. Execute the following query:3. 
```
MATCH (n0), (n1)-[r0 *..3]-(n2 :L2) WHERE ((n2.k13) <= (n2.k13)) WITH min('r') AS a0, n2, min('c9XMP') AS a1 WHERE ((n2.k18) <= 1590293959) OPTIONAL MATCH (n2) OPTIONAL MATCH (n2)-[r1 *]-(n3 :L5) WITH sum(-243320739) AS a5 where false RETURN 2
```
or
```
MATCH (n0), (n1)-[r0 *..3]-(n2 :L2) WHERE ((n2.k13) <= (n2.k13)) WITH min('r') AS a0, n2, min('c9XMP') AS a1 WHERE ((n2.k18) <= 1590293959) OPTIONAL MATCH (n2) OPTIONAL MATCH (n2)-[r1 *]-(n3 :L5) WITH max('') AS a3 WHERE false RETURN 2
```

# My discovery and assumption
The above two queries take more than a minute to execute in 4.4.10 while in 4.2.19 they only needs a few mini seconds. An interesting thing to notice is that if you delete the "where false" in above queries, any of them takes more than a minute in both 4.4.10 and 4.2.19.
So my assumption is that 4.2.19 has done some optimization to find that the last aggregation function is not required to be calculated since the last where clause is always false. However, for an unknown reason, started from 4.3, such optimization is disabled and thus cause unnecessary execution cost.
I wonder if this is done accidentally during upgrading or this is done on purpose. 
Another advice is that, since functions such as max('') has constant result and can only end up with one line if other columns are not considered. Can such functions be optimized too?

**Comment by HannesSandberg** [comment]
Hi! Thanks for your report. We will investigate this.

**Comment by pontusmelke** [comment]
@ddaa2000 Yes you are right we did change this a little bit in 4.3. For cases where we now at planning time that a predicate is false, such as `MATCH (n) WHERE 1 < 0 RETURN 0` we used to plan a `DropResult` operation that basically did nothing, in 4.3 we changed this to instead insert a `LIMIT 0` which in most cases is equivalent. However for your particular query where there is an eager aggregation after the LIMIT 0 we end up actually performing the operation even if we don't strictly need. 

We have created a fix and in upcoming versions performance should be on par with 4.2 for these types of queries.

Thanks for reporting!

Best regards,
Pontus
